### PR TITLE
Flexible dict

### PR DESCRIPTION
**Pull request by pszufe**
Introduction of flexible dictionaries including `SortedDict` for simulation results replicability.

### CODE REVIEW

File: docs/src/reference.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ h
 
 # output
 
-2×3 Hypergraph{Int64,Nothing,Nothing}:
+2×3 Hypergraph{Int64,Nothing,Nothing,Dict{Int64,Int64}}:
```
Comment (bkamins) [inline docs/src/reference.md]: maybe we should add `summary` method for `Hypergraph`?

File: src/hypergraph.jl [hunk 2]
```diff
@@ -114,6 +151,14 @@ If a vertex does not belong to a hyperedge `nothing` is returned.
     get(h.v2he[idx[1]], idx[2], nothing)
 end
 
+@inline function Base.pop!(d::SortedDict,key::Int,::Nothing)
```
Comment (bkamins) [inline src/hypergraph.jl]: this should be a PR do DataStructures.jl as doing it this way is type piracy

File: src/hypergraph.jl [hunk 3]
```diff
@@ -5,85 +5,122 @@ A hypergraph storing information about vertices and hyperedges.
...
+end
+
+function Hypergraph(m::AbstractMatrix{Union{T, Nothing}}) where {T<:Real}
+    Hypergraph{T,Nothing,Nothing,Dict{Int,T}}(m)
+end
+
 
 function Hypergraph(g::LightGraphs.Graph)
-    h = Hypergraph{Bool}(maximum(vertices(g)), ne(g))
+    h = Hypergraph{Bool,Nothing,Nothing,SortedDict{Int,Bool}}(maximum(vertices(g)), ne(g))
```
Comment (bkamins) [inline src/hypergraph.jl]: why do you use `SortedDict` here?

Comment (pszufe) [reply src/hypergraph.jl]: `LightGraphs.SimpleGraph` keeps the adjacency list in a deterministic sorted order. Hence the equivalent will be `SortedDict`

File: src/hypergraph.jl (hunk 2 above)
Comment (pszufe) [reply src/hypergraph.jl]: done: https://github.com/JuliaCollections/DataStructures.jl/pull/566 
Let's see if this gets accepted than we can drop that function from here.

**Comment by bkamins** [comment]
delete the branch from the repo when you are done with it pls.

**Pull request merged**